src/replication.c

```diff
@@ -1371,57 +1373,87 @@ check_for_command_completion(spec_t *spec, rcommon_cmd_t *rcomm_cmd, ISTGT_LU_CM
+	int success = 0, failure = 0, healthy_response = 0, response_received;
 	int min_response;
+	int healthy_replica = 0;
 
 	for (i = 0; i < rcomm_cmd->copies_sent; i++) {
-		if (rcomm_cmd->resp_list[i].status == 1) {
+		if (rcomm_cmd->resp_list[i].status & RECEIVED_OK) {
 			success++;
-		} else if (rcomm_cmd->resp_list[i].status == -1) {
+			if (rcomm_cmd->resp_list[i].status & SENT_TO_HEALTHY)
+				healthy_response++;
+		} else if (rcomm_cmd->resp_list[i].status & RECEIVED_ERR) {
 			failure++;
 		}
+
+		if (rcomm_cmd->resp_list[i].status & SENT_TO_HEALTHY)
+			healthy_replica++;
 	}
 
 	response_received = success + failure;
 	min_response = MAX(rcomm_cmd->replication_factor -
 	    rcomm_cmd->consistency_factor + 1, rcomm_cmd->consistency_factor);
 
+	rc = 0;
 	if (rcomm_cmd->opcode == ZVOL_OPCODE_READ) {
-		response_received = success + failure;
-
-		if (rcomm_cmd->copies_sent >= min_response) {
-			if (response_received < min_response) {
-				rc = 0;
-			} else {
-				if (failure >= min_response ||
-				    success < min_response) {
-					rc = -1;
-				} else {
-					data = handle_read_consistency(
-					    rcomm_cmd, spec->blocklen,
-					    true);
-					cmd->data = data;
-					rc = 1;
-				}
-			}
-		} else {
-			if (rcomm_cmd->copies_sent == success) {
-				data = handle_read_consistency(rcomm_cmd,
-				    spec->blocklen, false);
-				cmd->data = data;
-				rc = 1;
-			}
-
-			if (rcomm_cmd->copies_sent == failure)
-				rc = -1;
+		/*
+		 * If the command is sent to single replica only then
+		 * min_response should be set to 1.
+		 */
+		min_response = (rcomm_cmd->copies_sent == 1) ? 1 : min_response;
+		if (success >= min_response) {
+			/*
+			 * we got the successful response from the required
+			 * number of the replica. So, we can reply back to the
+			 * client with success
+			 */
+			data = handle_read_consistency(rcomm_cmd,
+			    spec->blocklen, (rcomm_cmd->copies_sent == 1) ?
+			    false : true);
+			cmd->data = data;
+			rc = 1;
+		} else if (response_received == rcomm_cmd->copies_sent) {
+			/*
+			 * we have received the response from all the replicas
+			 * to whom we have sent a command, but we didn't get
+			 * enough number of successful response. So, we will
+			 * reply to the client with an error.
+			 */
+			rc = -1;
+			REPLICA_ERRLOG("didn't receive success from replica.."
+			    " cmd:read io(%lu) cs(%d)\n", rcomm_cmd->io_seq,
+			    rcomm_cmd->copies_sent);
+		} else if ((rcomm_cmd->copies_sent - failure) < min_response) {
+			/*
+			 * In this case, we will avoid waiting for replicas
+			 * which haven't sent response yet.
+			 */
+			REPLICA_ERRLOG("got error from %d replicas.. io(%lu)"
+			    "cs(%d)\n", failure, rcomm_cmd->io_seq,
+			    rcomm_cmd->copies_sent);
+			rc = -1;
 		}
 	} else if (rcomm_cmd->opcode == ZVOL_OPCODE_WRITE) {
-		if (success >= min_response) {
+		if (healthy_replica >= rcomm_cmd->consistency_factor &&
+		    healthy_response >= rcomm_cmd->consistency_factor) {
```

Comment:
just checking for healthy_response should be good. healthy_replica check can be added as assert in debug builds.